deprecating array of args support in appium service

tamil777selvan opened the pull request:
## Proposed changes

deprecating array of args support in appium service 

## Types of changes

[//]: # (What types of changes does your code introduce to WebdriverIO?)
[//]: # (_Put an `x` in the boxes that apply_)

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Documentation update

## Checklist

[//]: # (_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._)

- [x] I have read the [CONTRIBUTING](https://github.com/webdriverio/webdriverio/blob/main/CONTRIBUTING.md) doc
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)
- [ ] I have added proper type definitions for new commands (if appropriate)

## Further comments

[//]: # (If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...)

### Reviewers: @webdriverio/project-committers

christian-bromann commented on packages/wdio-appium-service/src/utils.ts:
```diff
@@ -23,6 +23,36 @@ export function getFilePath (filePath: string, defaultFilename: string): string
     return absolutePath
 }
 
+export function defragCliArgs(args?: AppiumServerArguments | Array<string>) {
+    if (Array.isArray(args)) {
+        return args.reduce((acc: KeyValueArgs, currentItem, index, array) => {
+            const nextItem = array[index + 1]
+
+            if (
+                currentItem?.toString().startsWith('--') ||
+                currentItem?.toString().startsWith('-')
+            ) {
+                const [key, value] = (currentItem ?? '').toString().split('=')
+
+                if (value !== undefined) {
+                    acc[key.replace(/^-+/g, '')] = value
+                } else if (
+                    nextItem &&
+                    (nextItem?.toString().startsWith('--') ||
+                        nextItem?.toString().startsWith('-'))
+                ) {
+                    acc[key.replace(/^-+/g, '')] = true
+                } else {
+                    acc[key.replace(/^-+/g, '')] = nextItem
+                }
+            }
+            return acc
+        }, {})
+    }
+
+    return args
+}
```
Let's return early
```suggestion
export function defragCliArgs(args?: AppiumServerArguments | Array<string>) {
    if (!Array.isArray(args)) {
        return args
    }

    return args.reduce((acc: KeyValueArgs, currentItem, index, array) => {
        const nextItem = array[index + 1]

        if (
            !currentItem?.toString().startsWith('--') &&
            !currentItem?.toString().startsWith('-')
        ) {
            return acc
        }

        const [key, value] = (currentItem ?? '').toString().split('=')

        if (value !== undefined) {
            acc[key.replace(/^-+/g, '')] = value
        } else if (
            nextItem &&
            (nextItem?.toString().startsWith('--') ||
                nextItem?.toString().startsWith('-'))
        ) {
            acc[key.replace(/^-+/g, '')] = true
        } else {
            acc[key.replace(/^-+/g, '')] = nextItem
        }
        return acc
    }, {})
}
```

christian-bromann reviewed (commented):
I think it would be more clear if the `defragCliArgs` method could be incorporated into `formatCliArgs`. Also can we have more tests for the new function?

tamil777selvan commented:
> I think it would be more clear if the `defragCliArgs` method could be incorporated into `formatCliArgs`. Also can we have more tests for the new function?

At first, I considered changing the `formatCliArgs` function instead of creating a new one. However, because `_args` is used in `_setCapabilities`, which also needs the `port` value, I ended up needing this new method.

Also as suggested, I have updated the new method and added more tests to check it's functionality. Please share your thoughts on this.

christian-bromann commented:
@tamil777selvan I am a bit confused. It seems to be that with `defragCliArgs` we now support to pass in Appium args as object __and__ array. If this is the case I suggest to only go with one option since it can get confusing otherwise. E.g. update the docs and make sure check the type of the `_args` object in the service.

tamil777selvan commented:
> @tamil777selvan I am a bit confused. It seems to be that with `defragCliArgs` we now support to pass in Appium args as object **and** array. If this is the case I suggest to only go with one option since it can get confusing otherwise. E.g. update the docs and make sure check the type of the `_args` object in the service.

Hi @christian-bromann, The [documentation](https://webdriver.io/docs/appium-service#args) explains its capability to accept objects/arrays as inputs for 'args'. Could we consider discontinuing the array support in this scenario? Please share your opinion.

christian-bromann commented:
> Could we consider discontinuing the array support in this scenario? Please share your opinion.

Yeah, I would suggest that. Given array support never worked, we can safely discontinue it. It is less confusing to only support 1 type here. Would you mind making that change?

tamil777selvan commented:
> > Could we consider discontinuing the array support in this scenario? Please share your opinion.
> 
> Yeah, I would suggest that. Given array support never worked, we can safely discontinue it. It is less confusing to only support 1 type here. Would you mind making that change?

Yeah sure, on it.

tamil777selvan commented:
Hi @christian-bromann, I've eliminated the array support for arguments and enforced the type to be an object. Please share your opinion.

christian-bromann commented on packages/wdio-appium-service/README.md:
```diff
@@ -86,9 +86,9 @@ export const config = {
 
 See [the documentation](https://appium.github.io/appium.io/docs/en/writing-running-appium/server-args/index.html) for possible arguments.
-The arguments should be supplied in lower camel case, so `--pre-launch true` becomes `preLaunch: true` or passed as an array.
+The arguments are supplied in lower camel case. For instance, `debugLogSpacing: true` transforms into `--debug-log-spacing true`, or they can be supplied as outlined in the Appium documentation.
```
Boolean flags should not have a value.
```suggestion
The arguments are supplied in lower camel case. For instance, `debugLogSpacing: true` transforms into `--debug-log-spacing`, or they can be supplied as outlined in the Appium documentation.
```

Can you link _Appium documentation_ to an Appium resource that could be of help here?

tamil777selvan replied:
Furthermore, the documentation link is connected in line 88; however, it points to the outdated Appium version. Would it be advisable to directly link to the documentation [here](https://github.com/appium/appium/blob/master/packages/appium/docs/en/cli/args.md) on GitHub?

christian-bromann replied:
yes, that makes sense

christian-bromann commented on packages/wdio-appium-service/README.md:
```diff
@@ -102,10 +102,10 @@ export const config = {
                 // ...
                 debugLogSpacing: true,
                 platformName: 'iOS',
+                '-p': 1234,
+                '--relaxed-security': true
```
I don't think we should document such usage and encourage users to write out the full property name.

tamil777selvan replied:
Understood, but would this apply equally to aliases as well?

christian-bromann replied:
I am not sure if we should support these, I think if someone e.g. sets `{ p: 1234 }` we can create `['-p', 1234]` out of it. What do you think?

tamil777selvan replied:
I've observed that this approach isn't applicable to all aliases. 

For example: if I have `{ p: 1234 }`, following the formatting, it would result in `['--p', 1234]`, which Appium would accept.

However, consider a scenario where I provide an address in the alias, such as `{ a: 'foo' }`. The formatting would lead to `['--a', 'foo']`, triggering an ambiguous error in Appium.

To address this situation, it becomes necessary to either permit the use of full property names in the args or to consider discontinuing the support by updating the documentation. Please share your opinion.

christian-bromann replied:
From the user experience point of view I feel like this makes the configuration more clear anyway. What do you think?

tamil777selvan replied:
Certainly, that sounds like a solid plan. I'll integrate those adjustments and make the necessary updates to the documentation.

tamil777selvan replied:
Hi @christian-bromann, I have implemented all the required modifications. Please review them at your convenience.

christian-bromann commented on packages/wdio-appium-service/tests/launcher.test.ts:
```diff
@@ -388,6 +389,16 @@ describe('Appium launcher', () => {
+                args: []
+            }
+            const launcher = new AppiumLauncher(options, [], {} as any)
+            const error = await launcher.onPrepare().catch((err) => err)
+            const expectedError = new Error('Args should be an object')
+            expect(error).toEqual(expectedError)
```
```suggestion
            const expectedError = new Error('Args should be an object')
            await expect(launcher.onPrepare()).rejects.toEqual(expectedError)
```

christian-bromann commented on packages/wdio-appium-service/src/launcher.ts:
```diff
@@ -118,6 +118,13 @@ export default class AppiumLauncher implements Services.ServiceInstance {
+         * Throws an error if `this._options.args` is defined and is an array.
+         * @throws {Error} If `this._options.args` is an array.
+         */
+        if (this._options.args && Array.isArray(this._options.args)) {
```
this will be sufficient
```suggestion
        if (Array.isArray(this._options.args)) {
```

christian-bromann reviewed (commented):
Since we are already at this, could we extend the `AppiumServerArguments` and actually fill it up with given server properties rather than just having: `[capability: string]: any`. The rest looks good to me!

tamil777selvan commented:
> Since we are already at this, could we extend the `AppiumServerArguments` and actually fill it up with given server properties rather than just having: `[capability: string]: any`. The rest looks good to me!

I have included all the parameters and corresponding types utilised by the Appium server CLI. Please review them at your convenience.

christian-bromann reviewed (approved):
This is awesome, thank you so much 👍
